tiledb/sm/query/readers/aggregators/null_count_aggregator.cc

```diff
@@ -0,0 +1,158 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class NullCountAggregator.
+ */
+
+#include "tiledb/sm/query/readers/aggregators/null_count_aggregator.h"
+
+#include "tiledb/sm/query/query_buffer.h"
+#include "tiledb/sm/query/readers/aggregators/aggregate_buffer.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class NullCountAggregatorStatusException : public StatusException {
+ public:
+  explicit NullCountAggregatorStatusException(const std::string& message)
+      : StatusException("NullCountAggregator", message) {
+  }
+};
+
+template <typename T>
+NullCountAggregator<T>::NullCountAggregator(const FieldInfo field_info)
+    : field_info_(field_info)
+    , null_count_(0) {
+  if (!field_info_.is_nullable_) {
+    throw NullCountAggregatorStatusException(
+        "NullCount aggregates must only be requested for nullable attributes.");
+  }
+}
+
+template <typename T>
+void NullCountAggregator<T>::validate_output_buffer(
+    std::string output_field_name,
+    std::unordered_map<std::string, QueryBuffer>& buffers) {
+  if (buffers.count(output_field_name) == 0) {
+    throw NullCountAggregatorStatusException("Result buffer doesn't exist.");
+  }
+
+  auto& result_buffer = buffers[output_field_name];
+  if (result_buffer.buffer_ == nullptr) {
+    throw NullCountAggregatorStatusException(
+        "NullCount aggregates must have a fixed size buffer.");
+  }
+
+  if (result_buffer.buffer_var_ != nullptr) {
+    throw NullCountAggregatorStatusException(
+        "NullCount aggregates must not have a var buffer.");
+  }
+
+  if (result_buffer.original_buffer_size_ != 8) {
+    throw NullCountAggregatorStatusException(
+        "NullCount aggregates fixed size buffer should be for one element "
+        "only.");
+  }
+
+  bool exists_validity = result_buffer.validity_vector_.buffer();
+  if (exists_validity) {
+    throw NullCountAggregatorStatusException(
+        "NullCount aggregates must not have a validity buffer.");
+  }
+}
+
+template <typename T>
```

Comment:
I'm not sure we need this aggregator to be typed, I double checked below to see if we use `T` at all and it doesn't seem like we do. Am I missing something?

Reply:
Nice catch!